### PR TITLE
simplifying the quick_test and handling trivial cases first

### PR DESCRIPTION
**Pull request by mmatera**
Some smaller changes but have measurable effect over many symbols.

### CODE REVIEW

**Comment by mmatera** [comment]
imagen

imagen


imagen
imagen

**Comment by mmatera** [comment]
OK, this one seems to worth it...

**Comment by rocky** [comment]
> OK, this one seems to be worth...

Good work!  I'd like to look at in detail when I get a chance.

@mmatera @TiagoCavalcante have any idea what's up with the the bar chart lengths not matching the reported percentage differences. Like the one  -52.18%. The bar is clearly not less than half. 

If you have the actual time values, we should be able to figure out which is off. Or is this a logarithmic scale thing?

**Comment by mmatera** [comment]
> > OK, this one seems to be worth...
> 
> Good work! I'd like to look at in detail when I get a chance.
> 
> @mmatera @TiagoCavalcante have any idea what's up with the the bar chart lengths not matching the reported percentage differences. Like the one -52.18%. The bar is clearly not less than half.
> 
> If you have the actual time values, we should be able to figure out which is off. Or is this a logarithmic scale thing?
It is just the log scale...

**Comment by rocky** [comment]
@mmatera If you want to merge this it is fine with me. 

Sometime I would like to try to replicate the results to double check, and possibly understand what's up with the few small tests that are slower. 

However, I can do that after merge. It might be the weekend before I can get around to this  and it is not fair to block things in case you want to build off of this.

**Comment by rocky** [comment]
@mmatera If you are going to be pushing more commits like this, let's get more benchmarks in there or rerun the existing ones.

If this was motivated by something in particular we are in the dark as what was going on.

**Comment by mmatera** [comment]
@rocky, for some reason, the improvements I made were fragile when I merge it with the improvements for  `N`. I am trying to figure out what are the critical factors...

**Comment by rocky** [comment]
> @rocky, for some reason, the improvements I made were fragile when I merge it with the improvements for `N`. I am trying to figure out what are the critical factors...

Ok - thanks for the explanation.

**Comment by mmatera** [comment]
Some stupid changes, like 
```self.name in system_symbols("Pi","E", \ldots)
``` 
to 

```
name = self.name
any( name==s for s in  system_symbols("Pi","E", \ldots))
```

in the  `Symbol.is_numeric` method seems to have a measurable impact in the performance. As they are related to this PR, I want to get the best option before merging.

**Comment by TiagoCavalcante** [comment]
@mmatera what about `name in sytem_symbols(...)`?

**Comment by mmatera** [comment]
> @mmatera what about `name in sytem_symbols(...)` instead of the `any` approach?

It seems that to check if a string is equal to one in a list of other strings, the faster way is to put the candidates in a tuple (not a list, not a set) and then use `in`:
`name in tuple(candidate1, candidata2,...)`

methods like `is_numeric` are used intensively in the pattern matching code, so any speedup (or slowdown) here has a notorious impact over all the timings...

**Comment by mmatera** [comment]
Another bottleneck seems to be in the way `get_head_name()` is called:  each time you want to know if `a` or `7.2` is a number,
a call to `self.get_head_name()`  is done, that at the time calls `self.get_head()`, that at the time calls (for atoms)  `Symbol(self.get_atom_name())`
I had the idea that as this part is inside a Cythonized module, all these calls would be done inline. However, doing benchmarks over this it seems it is not the case.

**Comment by TiagoCavalcante** [comment]
Mabye use a class constant?

**Comment by rocky** [comment]
One thing we should be thinking about and I'd like to try to address when the PR is finishes it to make sure we describe why particular things have been done that way and  not  some other way.

In particular, there was a time, not too long ago, when code was getting converted into more sophisticated pattern rules.  Unless the reasons behind various actions are communicated (and that does not mean here in PR comments, but in the code and/or documentation somewhere) this knowledge which took a bit of time to figure out will be lost. This is pretty much the situation we were given. 

So for example if we've benchmarked the system and find the `N`  or other built-in functions are crucial for performance, then we should be noting that and providing data to show this. 

Personally, I tend to so this as I code, but we should make sure that we go over things so that before the merge things _describe_ what we want to describe. 

Programming is not _just_ about instructing a computer what to do, but also for communicating with other programmers what the problems, issues, and solutions are.

**Comment by TiagoCavalcante** [comment]
> Mabye use a class constant?

I have been looking at this, and this doesn't seems to be possible without a lot of work.

**Comment by rocky** [comment]
I haven't been following the details of what's going on except in a more superficial way. 

We don't have to do everything in one PR. If we need a class constant or some other structures, we can add that later.

**Review by TiagoCavalcante — COMMENTED** [review]
Thanks for doing this, the improvements seems to be in every place.

File: mathics/core/expression.py [hunk 1]
```diff
@@ -327,7 +327,9 @@ def get_lookup_name(self):
     def get_head(self):
         return None
 
-    def get_head_name(self):
+    def get_head_name(self) -> "str":
+        print("get_head_name of ", self)
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/core/expression.py:331]: Iit isjust for debugging. This method shouldn't be ever called

File: mathics/core/expression.py [hunk 2]
```diff
@@ -327,7 +327,9 @@ def get_lookup_name(self):
     def get_head(self):
         return None
 
-    def get_head_name(self):
+    def get_head_name(self) -> "str":
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/core/expression.py:330]: Yep

File: mathics/core/expression.py [hunk 3]
```diff
@@ -2069,7 +2072,7 @@ def equal2(self, rhs: Any) -> Optional[bool]:
 
     def sameQ(self, rhs: Any) -> bool:
         """Mathics SameQ"""
-        return id(self) == id(rhs) or isinstance(rhs, Symbol) and self.name == rhs.name
+        return isinstance(rhs, Symbol) and self.name == rhs.name
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/core/expression.py:2075]: It tries to be both, at the light of what I understand now about Python

File: mathics/core/expression.py [hunk 4]
```diff
@@ -2112,22 +2127,50 @@ def __getnewargs__(self):
...
+SymbolPi = Symbol("Pi")
 SymbolRule = Symbol("Rule")
 SymbolSequence = Symbol("Sequence")
+SymbolStringQ = Symbol("StringQ")
 SymbolTrue = Symbol("True")
 SymbolUndefined = Symbol("Undefined")
-SymbolLess = Symbol("Less")
-SymbolGreater = Symbol("Greater")
+
+arithmetic_head_symbols = system_symbols(
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/core/expression.py:2155]: This is just for simplicity.

File: mathics/core/expression.py [hunk 5]
```diff
@@ -2202,6 +2245,7 @@ def _NumberFormat(man, base, exp, options):
 
 class Integer(Number):
     value: int
+    class_head_name = "System`Integer"
```
[the remark this answers is not in the document]
Comment (mmatera) [reply mathics/core/expression.py:2248]: It is not used yet. I was trying to check your idea, but I couldn't do it today

**Comment by mmatera** [comment]
> If you are seeing a significant difference here, I am beginning to wonder...

The problem with these routines is that we use them each time we compare two expressions (and I think we do each time we do pattern matching):


This is the timing for the plane call of the three comparisons called in `Symbol.sameQ`
image


This is for the Cython compiled version:

image


So, maybe it makes sense to keep this comparison in  `Expression.sameQ` but not in `Symbol.sameQ`.

**Pull request closed without merging**